_posts/2022-11-11-software-factory.markdown

```diff
@@ -0,0 +1,127 @@
+* give feedbacks on what's going wrong during builds
+
+All of the above goals can be achieved defining build profiles and flexible agents, backed up with our own set of internal functions (called _Jenkins Shared Library_ or _JSL_), encouraging and, in particular projects, enforcing to address critical issues discovered in our software supply chain.
+
+The _JSL_ simplifies functionalities for building, testing and deploying applications. The developers do not have to make any adjustments to their pipelines when the environment or default settings for functions change, as these can be made globally for everyone in the JSL. Furthermore, we have clustered the JSL functions according to functionalities such as git, jira, maven, gradle, nexus, trivy, docker, and others. For example, the Maven functions offer simple options for querying the version of the current Jira project, just with the Jira project ID, and setting it in the Maven project:
+
+```groovy
+@Library('gematik-jenkins-shared-library') _
+def JIRA_PROJECT_ID = "..."
+pipeline {
+    ...
+    stages {
+        ...
+        stage('set Version from Jira-Project') {
+            steps {
+                mavenSetVersionFromJiraProject(JIRA_PROJECT_ID)
+            }
+        }
+    }
+    ...
+}
+```
+
+Another example are the vulnerability scan functionalities for Docker images, with one command an HTML report is created for an image and saved as a result for the pipeline:
+
+```groovy
+@Library('gematik-jenkins-shared-library') _
+def IMAGE_NAME= "MyImage"
+def VERSION = "latest"
+
+pipeline {
+    ...
+    stages {
+        ...
+        stage('Step') {
+            steps {
+                trivyVulnerabilitiesScanAllAsHtml(IMAGE_NAME, VERSION)
+            }
+        }
+    }
+    ...
+}
+```
+
+Before hitting the production, artifacts are properly scanned for security or misconfiguration issues and tested using [our own test-suites]({{site.baseurl}}/testing/2022/10/13/zeroline-test-suite). If problems are found, they are visible in form of visual feedback from Jenkins and GitLab but also in form of chat notifications through Microsoft Teams.
+
+Another example of friction-less usage of the Software Factory can be proven by the following Scenario: imagine you want to publish an internally built Docker Image to Docker Hub. The JSL not only contains individual functionalities, but also complete pipelines. 
+The following Jenkins Pipeline use the gematik Template Pipeline _pipelineDockerHubPublish_ and need only 2 Parameters:
+
+```groovy
+@Library('gematik-jenkins-shared-library') _
+pipelineDockerHubPublish {
+    INTERNAL_IMAGE_NAME = "demis/hospital-location-service"
+    EXTERNAL_IMAGE_NAME = "gematik1/demis-hospital-location-service"
+}
+```
+
+which then translates to following Job:
+
+![Sketch of the matrix organization @gematik]({{site.baseurl}}/assets/img/221111-swfactory/pipeline.webp)
+*Example of pipeline resulting from Groovy source code*
+
+That's it!
```

Comment:
Maybe...or you add a paragraph on learnings and feedbacks from the teams :)
It would also be nice to have a (visual?) overview of the tools that are in use - perhaps even with a link to their websites.

Reply:
I've added the final words with the feedback part. @outbreaker is finalizing the visual overview, but I think we're good to go!